File: Solutions/MoNaDDu/Game Shop/4/Part4.hs
module Part4 where

import Control.Monad
import Control.Applicative
import Data.Char

data Item = Item { name :: String, description :: String, weight :: Int, value :: Int } |
            Weapon { name :: String, description :: String, weight :: Int, value :: Int, damage :: Int } |
            Armor { name :: String, description :: String, weight :: Int, value :: Int, defense :: Int } |
            Potion { name :: String, description :: String, weight :: Int, value :: Int, potionType :: PotionType, capacity :: Int }
 deriving ( Show )

data PotionType = HP | MP

instance Show PotionType where
    show HP = "Health"
    show MP = "Mana"

data Shop = Shop { shopName :: String, shopItems :: Items }

type Items = [Item]
type Shops = [Shop]

-- main 함수
main :: IO ()
main = do
    shops <- loadShops shopInformation
    selectShop shops

getInt :: IO Int
getInt = fmap read getLine :: IO Int

selectShop :: Shops -> IO ()
selectShop shops = do
    putStrLn "- Shop Select -"
    putStrLn "   1. Weapon/Armor Shop"
    putStrLn "   2. Potion Shop"
    putStrLn "   3. Exit\n"
    putStr "Select : "
    input <- getInt
    putStrLn ""
    let shopCount = 1 + length shops
    if input == shopCount then return () else do
            if input < shopCount then playShop (shops !! (input - 1)) else putStrLn "Invalid number! Try again.\n"
            selectShop shops

playShop :: Shop -> IO ()
playShop s = do
    let welcomeMessage = "Welcome to " ++ shopName s ++ " Shop!"
    putStrLn welcomeMessage
    replicateM_ (length welcomeMessage) (putStr "-")
    putStrLn ""
    putStrLn "   1. Show Item List"
    putStrLn "   2. Exit\n"
    putStr "Select : "
    input <- getInt
    putStrLn ""
    if input == 2 then return () else do
        case input of
            1 -> showItemList s
            _ -> putStrLn "Invalid number! Try again.\n"
        playShop s

-- Item 관련 함수
describe :: Items -> IO ()
describe = mapM_ (putStrLn . toString)

toString :: Item -> String
toString (Item name description weight value) = baseInfo (Item name description weight value)
toString (Weapon name description weight value damage) = baseInfo (Item name description weight value) ++
                                                         "Damage      = " ++ show damage ++ "\n"
toString (Armor name description weight value defense) = baseInfo (Item name description weight value) ++
                                                         "Defense     = " ++ show defense ++ "\n"
toString (Potion name description weight value potionType capacity) = baseInfo (Item name description weight value) ++
                                                                      "Type         = " ++ show potionType ++ "\n" ++
                                                                      "Capacity     = " ++ show capacity ++ "\n"

baseInfo :: Item -> String
baseInfo item = "Name        = " ++ name item ++ "\n" ++
                "Description = " ++ description item ++ "\n" ++
                "Weight      = " ++ w'' ++ "\n" ++
                "value       = " ++ v'' ++ "\n"
                where 
                w = weight item
                v = value item
                w' = show w
                v' = show v
                w''
                 | w > 1 = w' ++ " lbs"
                 | otherwise = w' ++ " lb"
                v''
                 | v > 1 = v' ++ " gold coins"
                 | otherwise = v' ++ " gold coin"

-- Shop 관련 함수
showItemList :: Shop -> IO ()
showItemList shop = do
    putStrLn $ "Welcom to " ++ shopName shop ++ "!\n"
    describe $ shopItems shop

-- Basic Parser
newtype Parser a = P (String -> [(a, String)])

instance Functor Parser where
    -- fmap :: (a -> b) -> Parser a -> Parser b
    fmap g p = P (\inp -> case parse p inp of
        [] -> []
        [(v, out)] -> [(g v, out)])

instance Applicative Parser where
    -- pure :: a -> Parser a
    pure v = P (\inp -> [(v, inp)])

    -- <*> :: Parser (a -> b) -> Parser a -> Parser b
    pg <*> px = P (\inp -> case parse pg inp of
        [] -> []
        [(g, out)] -> parse (fmap g px) out)

instance Monad Parser where
    -- (>>=) :: Parser a -> (a -> Parser b) -> Parser b
    p >>= f = P (\inp -> case parse p inp of
        [] -> []
        [(v, out)] -> parse (f v) out)        

instance Alternative Parser where
    -- empty :: Parser a
    empty = P (const [])

    -- (<|>) :: Parser a -> Parser a -> Parser a
    p <|> q = P (\inp -> case parse p inp of
        [] -> parse q inp
        [(v, out)] -> [(v, out)])

parse :: Parser a -> String -> [(a, String)]
parse (P p) = p

item :: Parser Char
item = P (\inp -> case inp of
    [] -> []
    (x:xs) -> [(x,xs)])

sat :: (Char -> Bool) -> Parser Char
sat p = do
    x <- item
    if p x then return x else empty

digit :: Parser Char
digit = sat isDigit

char :: Char -> Parser Char
char x = sat (== x)

string :: String -> Parser String
string [] = return []
string (x:xs) = do
    char x
    string xs
    return (x:xs)

nat :: Parser Int
nat = do
    xs <- some digit
    return (read xs)

space :: Parser ()
space = do
    many (sat isSpace)
    return ()

int :: Parser Int
int = do
    char '-'
    n <- nat
    return (-n)
    <|> nat

token :: Parser a -> Parser a
token p = do
    space
    v <- p
    space
    return v

integer :: Parser Int
integer = token int

symbol :: String -> Parser String
symbol xs = token (string xs)

-- Element 데이터 타입 선언
data Element = S String | I Int
    deriving (Eq, Ord, Show, Read)

elementToString :: Element -> String
elementToString (S x) = x
elementToString (I x) = show x

elementToInt :: Element -> Int
elementToInt (I x) = x
elementToInt (S x) = read x

-- CSV Parser
parseCSV :: String -> [Element]
parseCSV xs = case parse csv xs of
    [(n, [])] -> n
    [(_, out)] -> error ("Unused input " ++ out)
    [] -> error "Invalid input"

-- 문법
-- csv ::= factor (, factor | e)
-- factor ::= quotes | int
-- integer ::= ... | -2 | -1 | 0 | 1 | 2 | ...
csv :: Parser [Element]
csv = do
    n <- factor
    ns <- many (do
        symbol ","
        factor)
    return (n:ns)

factor :: Parser Element
factor = do
    xs <- quotes
    return (S xs)
    <|> do
        ns <- integer
        return (I ns)

quotes :: Parser String
quotes = do
    symbol "\""
    xs <- many $ sat (/= '\"')
    symbol "\""
    return xs

-- 파일 입출력
readLines :: FilePath -> IO [String]
readLines = fmap lines . readFile

readCSV :: FilePath -> IO [[Element]]
readCSV fileName = do
    ls <- readLines fileName
    let results = map parseCSV ls
    return results

makeShop :: String -> Items -> Shop
makeShop name items = Shop { shopName = name, shopItems = items }

makeItems :: [[Element]] -> Items
makeItems = map makeItem

makeItem :: [Element] -> Item
makeItem [] = undefined 
makeItem (x:xs)
 | len >= 5 && x == S "Weapon" = Weapon { name = elementToString $ head xs, description = elementToString $ xs !! 1, weight = elementToInt $ xs !! 2, value = elementToInt $ xs !! 3, damage = elementToInt $ xs !! 4 }
 | len >= 5 && x == S "Armor" = Armor  { name = elementToString $ head xs, description = elementToString $ xs !! 1, weight = elementToInt $ xs !! 2, value = elementToInt $ xs !! 3, defense = elementToInt $ xs !! 4 }
 | len >= 6 && x == S "Potion" = Potion  { name = elementToString $ head xs, description = elementToString $ xs !! 1, weight = elementToInt $ xs !! 2, value = elementToInt $ xs !! 3, potionType = makePotionType $ elementToString $ xs !! 4, capacity = elementToInt $ xs !! 5 }
 | otherwise = error "Invalid item"
 where len = length xs

shopInformation = [
    ("Weapon/Armor", "equip_item.txt"),
    ("Potion", "potion_item.txt")
    ]

loadShops :: [(String, FilePath)] -> IO Shops
loadShops = mapM loadShop

loadShop :: (String, FilePath) -> IO Shop
loadShop (shopName, fileName) = do
    elements <- readCSV fileName
    let result = makeShop shopName $ makeItems elements
    return result

makePotionType :: String -> PotionType
makePotionType s
    | s == "Health" = HP
    | s == "Mana" = MP
    | otherwise = error "Invalid string"

File: Solutions/MoNaDDu/Game Shop/4/Part4.cpp
#include <iostream>
#include <string>
#include <iomanip>
#include <vector>
#include <memory>
#include <unordered_map>
#include <algorithm>
#include <fstream>
#include <regex>
#include <iterator>

using namespace std;

class Item;
class Weapon;
class Armor;
class Potion;
class Shop;
class ShopManager;

using ItemPtr = unique_ptr<Item>;
using WeaponPtr = unique_ptr<Weapon>;
using ArmorPtr = unique_ptr<Armor>;
using PotionPtr = unique_ptr<Potion>;
using ShopPtr = unique_ptr<Shop>;
using ShopManagerPtr = unique_ptr<ShopManager>;

class Item
{
	friend istream& operator>>(istream& is, ItemPtr& item);
public:
	Item() = default;

	Item(const string& name, const string& description, const int weight, const int value)
		: name(name), description(description), weight(weight), value(value)
	{
		
	}

	virtual ~Item() = default;

	virtual void Describe(void) const
	{
		DescribeName();
		DescribeDescription();
		DescribeWeight();
		DescribeValue();
	}

protected:
	static int GetSpaceInfo(void)
	{
		return space;
	}

private:
	string name;
	string description;
	int weight;
	int value;
	static const int space = 12;
	static const regex reg;

	inline void DescribeName(void) const
	{
		cout << left << setw(space) << "Name" << "= " << name << endl;
	}

	inline void DescribeDescription(void) const
	{
		cout << left << setw(space) << "Description" << "= " << description << endl;
	}

	inline void DescribeWeight(void) const
	{
		cout << left << setw(space) << "Weight" << "= " << weight << (weight > 1 ? " lbs" : " lb") << endl;
	}

	inline void DescribeValue(void) const
	{
		cout << left << setw(space) << "Value" << "= " << value << (value > 1 ? " gold coins" : " gold coin") << endl;
	}
};

const regex Item::reg(R"~([[:s:]"]*([^",]+)[[:s:]]*)~");

class Weapon : public Item
{
public:
	static ItemPtr MakeInstance(const string& name, const string& description, const int weight, const int value, const int damage)
	{
		return make_unique<Weapon>(name, description, weight, value, damage);
	}

	Weapon(const string& name, const string& description, const int weight, const int value, const int damage)
		: Item(name, description, weight, value), damage(damage)
	{

	}

	virtual ~Weapon() = default;

	virtual void Describe(void) const override
	{
		Item::Describe();
		DescribeDamage();
	}

private:
	int damage;

	inline void DescribeDamage(void) const
	{
		cout << setw(GetSpaceInfo()) << "Damage" << "= " << damage << endl;
	}
};

class Armor : public Item
{
public:
	static ItemPtr MakeInstance(const string& name, const string& description, const int weight, const int value, const int defense)
	{
		return make_unique<Armor>(name, description, weight, value, defense);
	}

	Armor(const string& name, const string& description, const int weight, const int value, const int defense)
		: Item(name, description, weight, value), defense(defense)
	{

	}

	virtual ~Armor() = default;

	virtual void Describe(void) const override
	{
		Item::Describe();
		DescribeDefense();
	}

private:
	int defense;

	inline void DescribeDefense(void) const
	{
		cout << setw(GetSpaceInfo()) << "Defense" << "= " << defense << endl;
	}
};

class Potion : public Item
{
public:
	enum class Type
	{
		HP,
		MP
	};

	static const unordered_map<string, Type> string_to_type;

	static ItemPtr MakeInstance(const string& name, const string& description, const int weight, const int value, const Type type, const int capacity)
	{
		return make_unique<Potion>(name, description, weight, value, type, capacity);
	}

	Potion(const string& name, const string& description, const int weight, const int value, const Type type, const int capacity)
		: Item(name, description, weight, value), type(type), capacity(capacity)
	{

	}

	virtual void Describe(void) const override
	{
		Item::Describe();
		DescribeType();
		DescribeCapacity();
	}

private:
	const int capacity;
	const Type type;
	static const unordered_map<Type, string> type_to_string;

	inline void DescribeType(void) const
	{
		cout << setw(GetSpaceInfo()) << "Type" << "= " << type_to_string.at(type) << endl;
	}

	inline void DescribeCapacity(void) const
	{
		cout << setw(GetSpaceInfo()) << "Capacity" << "= " << capacity << endl;
	}
};

// C++ 기본 라이브러리에는 bimap이 없어서 안타깝네요~
const unordered_map<Potion::Type, string> Potion::type_to_string{
	make_pair(Potion::Type::HP, "Health"),
	make_pair(Potion::Type::MP, "Mana")
};

const unordered_map<string, Potion::Type> Potion::string_to_type{
	make_pair("Health", Potion::Type::HP),
	make_pair("Mana", Potion::Type::MP)
};

class Shop
{
public:
	static ShopPtr MakeInstance(const string& name, vector<ItemPtr>&& items)
	{
		return make_unique<Shop>(name, move(items));
	}

	static ShopPtr MakeInstance(const string& name, const string& file_name)
	{
		return make_unique<Shop>(name, file_name);
	}

	Shop(const string& name, vector<ItemPtr>&& items)
		: name(name), items(move(items))
	{

	}

	Shop(const string& name, const string& file_name)
		: name(name)
	{
		ifstream istrm(file_name);
		if (!istrm.is_open())
			throw std::invalid_argument("Failed to open "s + file_name);
		else
		{
			while (!istrm.eof())
			{
				ItemPtr item;
				istrm >> item;	// operator>>를 오버로딩하여 입력처리 부분을 모듈화 함
				items.emplace_back(move(item));
			}
		}
	}

	virtual ~Shop() = default;

	void EmplaceBack(ItemPtr&& item)
	{
		items.emplace_back(move(item));
	}

	void ShowItemList(void) const
	{
		cout << "\n- Item List -\n";

		for (const auto& item : items)
		{
			item->Describe();
			cout << endl;
		}
	}

	const string& GetName(void) const
	{
		return name;
	}

private:
	const string name;
	vector<ItemPtr> items;
};

class ShopManager
{
public:
	static ShopManagerPtr MakeInstance(void)
	{
		return make_unique<ShopManager>();
	}

	void EmplaceBack(ShopPtr&& shop)
	{
		shop_list.emplace_back(move(shop));
	}

	template<typename F>
	void PrintShopName(F Func) const
	{
		for (const auto& name : shop_list)
			Func(name->GetName());
	}

	const string& GetShopName(const size_t index) const
	{
		return shop_list[index]->GetName();
	}

	void ShowItemListOfShop(const size_t index)
	{
		shop_list[index]->ShowItemList();
	}

	size_t GetNumberOfShop(void) const
	{
		return shop_list.size();
	}

private:
	vector<ShopPtr> shop_list;
};

ShopManagerPtr InitShopManager(void)
{
	vector<pair<string, string>> shops_information;

	shops_information.emplace_back(make_pair("Weapon/Armor", R"~(C:\Users\NaDDu\Documents\GitHub\ProgrammingPractice\Solutions\MoNaDDu\Game Shop\4\equip_item.txt)~"));
	shops_information.emplace_back(make_pair("Potion", R"~(C:\Users\NaDDu\Documents\GitHub\ProgrammingPractice\Solutions\MoNaDDu\Game Shop\4\potion_item.txt)~"));

	auto shop_manager = ShopManager::MakeInstance();

	try
	{
		for (const auto& shop_information : shops_information)
			shop_manager->EmplaceBack(Shop::MakeInstance(shop_information.first, shop_information.second));
	}
	catch (const exception& e)
	{
		cerr << e.what() << endl;
	}

	return shop_manager;
}

istream& operator>>(istream& is, ItemPtr& item)
{
	string input;

	getline(is, input);

	smatch m;
	vector<string> parsing_result;

	while (regex_search(input, m, Item::reg))
	{
		parsing_result.push_back(m[1]);
		input = m.suffix().str();
	}

	if (parsing_result.front() == "Weapon"s)
	{
		if (parsing_result.size() < 6)
			throw std::runtime_error("Not enough data.");

		item.reset(new Weapon(parsing_result[1], parsing_result[2], stoi(parsing_result[3]), stoi(parsing_result[4]), stoi(parsing_result[5])));
	}
	else if (parsing_result.front() == "Armor"s)
	{
		if (parsing_result.size() < 6)
			throw std::runtime_error("Not enough data.");

		item.reset(new Armor(parsing_result[1], parsing_result[2], stoi(parsing_result[3]), stoi(parsing_result[4]), stoi(parsing_result[5])));
	}
	else if (parsing_result.front() == "Potion"s)
	{
		if (parsing_result.size() < 7)
			throw std::runtime_error("Not enough data.");

		item.reset(new Potion(parsing_result[1], parsing_result[2], stoi(parsing_result[3]), stoi(parsing_result[4]), Potion::string_to_type.at(parsing_result[5]), stoi(parsing_result[6])));
	}
	else
		throw std::runtime_error("Unknown Item.");

	return is;
}

class MyGame
{
public:
	MyGame() = default;

	MyGame(ShopManagerPtr&& shop_manager)
		: shop_manager(move(shop_manager))
	{

	}

	~MyGame() = default;

	void SetShopManager(ShopManagerPtr&& shop_manager)
	{
		this->shop_manager = move(shop_manager);
	}

	int Play(void) const
	{
		PrintMenu();

		const auto selected_number = Input();
		auto end = false;

		if (selected_number < 0 || selected_number > shop_manager->GetNumberOfShop())
			cout << "\nInvalid number! Try again.\n" << endl;
		else if (selected_number != shop_manager->GetNumberOfShop())
			Shopping(selected_number);
		else
			end = true;

		return end ? 0 : Play();
	}

private:
	ShopManagerPtr shop_manager;

	size_t Input(void) const
	{
		size_t selected_number;

		cin >> selected_number;

		return selected_number - 1;
	}

	void Shopping(const size_t shop_number) const
	{
		auto shopping = true;

		while (shopping)
		{
			PrintShopMenu(shop_manager->GetShopName(shop_number));

			auto selected_shop_number = Input();

			switch (selected_shop_number)
			{
			case 0:
				shop_manager->ShowItemListOfShop(shop_number);
				break;
			case 1:
				shopping = false;
				break;
			default:
				cout << "\nInvalid number! Try again.\n" << endl;
				break;
			}
		}
	}

	void PrintMenu(void) const
	{
		auto index = 0u;

		cout << " - Shop Select -\n";
		shop_manager->PrintShopName([&index](const auto& name) {
			cout << "    " << ++index << ". " << name << " Shop\n";
		});
		cout << "    " << ++index << ". Exit\n\n";
		cout << "Select : ";
	}

	void PrintShopMenu(const string& shop_name) const
	{
		string welcome_message = "Welcome to " + shop_name + " Shop!";

		cout << '\n' << welcome_message << endl;
		fill_n(ostream_iterator<char>(cout), welcome_message.size(), '-');
		cout << "\n1. Show Item List\n";
		cout << "2. Exit\n\n";
		cout << "Select : ";
	}
};

int main()
{
	return MyGame(InitShopManager()).Play();
}